Source/WebCore/platform/graphics/gstreamer/mse/MediaPlayerPrivateGStreamerMSE.cpp

```diff
@@ -303,6 +321,12 @@ void MediaPlayerPrivateGStreamerMSE::updateStates()
             GST_ERROR_OBJECT(pipeline(), "Setting the pipeline to PAUSED failed");
         m_isPipelinePlaying = false;
     }
+
+    if (getStateResult == GST_STATE_CHANGE_SUCCESS && state >= GST_STATE_PAUSED) {
+        // Update playback rate that was set before the pipeline was prerolled.
+        // FIXME: This needs further investigation in https://bugs.webkit.org/show_bug.cgi?id=259776.
+        updatePlaybackRate();
+    }
```

Comment:
I've dig a bit further, I found this exact condition in `MediaPlayerPrivateGStreamer::updateStates()`, so I would adventure it may have been copied verbatim:

    if (getStateResult == GST_STATE_CHANGE_SUCCESS && m_currentState >= GST_STATE_PAUSED) {
        updatePlaybackRate();

Now, what does `updatePlaybackRate()` do?

1. Returns early if `!m_isChangingRate`.
2. Logs about the rate change about to occur.
3. Decides whether to mute audio depending on the rate set.
4. If `m_lastPlaybackRate != m_playbackRate`, calls `doSeek(playbackPosition(), m_playbackRate)`. Both regular and MSE implementations of `doSeek()` would rewrite a playback rate of 0 to 1 here, were it to occur. I suspect the goal here is to not do a seek if pausing/unpausing with rates != 0?
5. If `m_isPlaybackRatePaused` is set, it will be consumed (set to false) and the pipeline will be set to PLAYING if it wasn't already. **Therefore, there is therefore an implicit assumption that `updatePlaybackRate()` must not be called when `m_playbackRate == 0 && m_isPlaybackRatePaused`.** Indeed `MediaPlayerPrivateGStreamer::setRate()` will return early on rate == 0 to avoid that, just after setting the pipeline to PAUSED.
6. `m_isChangingRate` is reset to FALSE and `m_player->rateChanged()` is called, which will update the `m_reportedPlaybackRate` of the HTMLMediaElement with whatever `effectiveRate()` returns, which defaults to whatever `rate()` returns.

All this rate handling code looks quite messy to me. It's very hard for me to analyze. State changes are happening from many different places.

One goal with my MediaPlayerPrivateGStreamerMSE seek rework was cleaning up and simplifying it, which can be clearly seen by comparing `MediaPlayerPrivateGStreamerMSE::updateStates()` to `MediaPlayerPrivateGStreamer::updateStates()`.

I think the rate handling code could use a cleanup and/or refactor.

Reply:
After your analysis, surely you will correct me if I am wrong, I think the best case is keeping this but ensuring we are going going though that if playback is stopped because of zero rate. I updated the PR with that.